Contrib: Mac: Fix mac deploy so that the dmg background works

djp3 opened the pull request:
I couldn't get the mac .dmg build to work unless I made the two changes to python in this patch.
Update: I subsequently added a fix to the dmg build to make the background image work

djp3 committed:
Two changes to bytes and string management to get the mac deploy to work

kallewoof commented:
I ran into the `background.tiff` issue mentioned in https://github.com/bitcoin/bitcoin/issues/8120 when  I tried to check this.
https://github.com/theuni/bitcoin/commit/c622254170d4bd8c5ca75077214105cef666f4c4 addresses the encode/decode stuff in this PR as well but uses `output.decode` instead of `str(output, "utf-8")`.

fanquake commented:
Indeed, I wondered if this was going to sort out the background image issues. Lets try and fix those up here as well.

djp3 commented:
What is the background image issue?  I can probably fix that

kallewoof commented:
```Bash
$ make deploy
/Library/Developer/CommandLineTools/usr/bin/make -C src qt/bitcoin-qt
make[2]: `libsecp256k1.la' is up to date.
build-aux/install-sh -c -d Bitcoin-Qt.app/Contents/MacOS
STRIPPROG="/usr/bin/strip" /bin/sh /Users/me/workspace/bitcoin-kw/build-aux/install-sh -c -s  src/qt/bitcoin-qt Bitcoin-Qt.app/Contents/MacOS/Bitcoin-Qt
/usr/local/bin/python3.5 ./contrib/macdeploy/macdeployqtplus Bitcoin-Qt.app -add-qt-tr da,de,es,hu,ru,uk,zh_CN,zh_TW -translations-dir= -dmg -fancy ./contrib/macdeploy/fancy.plist -verbose 2 -volname Bitcoin-Core
Error: Could not find background picture at "background.tiff" or "./contrib/macdeploy/background.tiff"
make: *** [Bitcoin-Core.dmg] Error 1
```

djp3 commented:
`output.decode("utf-8")` and `str(output,"utf-8")` are equivalent according to the documentation:

"_If at least one of encoding or errors is given, object should be a bytes-like object (e.g. bytes or bytearray). In this case, if object is a bytes (or bytearray) object, then str(bytes, encoding, errors) is equivalent to bytes.decode(encoding, errors). Otherwise, the bytes object underlying the buffer object is obtained before calling bytes.decode(). See Binary Sequence Types — bytes, bytearray, memoryview and Buffer Protocol for information on buffer objects._" cite: [https://docs.python.org/3/library/stdtypes.html](https://docs.python.org/3/library/stdtypes.html)

djp3 commented:
Okay, I'll take a look at background.tiff issue in the morning.

jonasschnelli commented:
AFAIK, `make deploy` on native macOS machines is currently broken (introduced in #7192). Before 7192, there was just a tiff that was used as DMG background.
Since #7192, the tiff gets created by adding the package-name (`Bitcoin-Core`) on top of the `background.svg` file.

I think most devs compile releases over Gitian, so its not among the top priorities to fix. But we should.

djp3 committed:
Remove .svg image from mac build

djp3 committed:
Add new .jpg background image

djp3 committed:
Change parameters of configuration file for making .dmg build to accommodate new image background

djp3 committed:
Fix bug in dmg builder so that it actually reads in the configuration file

djp3 committed:
Change mac build process to use a static jpg background on the dmg rather than the svg build process

djp3 committed:
Removing build code that was for OSX machines that weren’t building for DARWIN.

djp3 commented:
Ok.  I changed the dmg build process to just use a static jpg image as the background.  I don't know why a dynamic background image was being built from an svg.  That seems kind of crazy.  In the process I found a bug in the macqtdeployplus script which I also fixed.

djp3 commented:
Checking on why it didn't pass tests

djp3 committed:
Undid the Makefile.am changes to try and get it through the travis tests

djp3 commented:
Okay, so if someone else can check this out on their machine.  There are some fixes to the python in the build process and I fixed the dmg background image.  It works fine on my machine and in the automated builds.

jonasschnelli commented:
I think this solution is a step backward. In #7192 we got rid of the static image and introduced a SVG with a flexible product name.

MarcoFalke commented:
> use a static jpg image as the background

Please don't use jpg for compression of such data. I think #9412 is the solution to go.

fanquake commented:
Agree with @jonasschnelli's comments.

Thanks for the initial bugfix and further contributions @djp3, however I'm going to close this in favour of #9412. It'd be great if you could test that PR and check it fixes your original issue.

djp3 commented on contrib/macdeploy/macdeployqtplus:
```diff
@@ -852,7 +852,7 @@ if config.dmg is not None:
             }
         if "window_bounds" in fancy:
-            params["window.bounds"] = ",".join([str(p) for p in fancy["window_bounds"]])
+            params["window_bounds"] = ",".join([str(p) for p in fancy["window_bounds"]])
```
@fanquake This wasn't in #9412

djp3 commented:
@fanquake #9412 worked for me.   There was one bug in this PR that wasn't in 9412 though. I'll tag you in the body of the code.  Should I generate a PR just for that?

fanquake commented:
@djp3 Thanks for pointing that out, a new PR won't be necessary, we can cherry pick the bug fix into #9412.

djp3 commented:
@fanquake Ok thanks!
